trusted_applet/internal/storage/slots/journal.go

```diff
@@ -0,0 +1,312 @@
+		Revision:   j.current.Revision + 1,
+		DataLen:    uint64(len(data)),
+		DataSHA256: h,
+		Data:       data,
+	}
+
+	cap := (j.start + j.length - j.nextBlock) * j.dev.BlockSize()
+	if cap < uint(e.Size()) {
+		// The record won't fit in the remaining space, so wrap around and write at the beginning.
+		j.nextBlock = j.start
+	}
+	// TODO(al): consider making this more "streamy".
+	buf := &bytes.Buffer{}
+	if err := marshalEntry(e, buf); err != nil {
+		return fmt.Errorf("failed to marshal entry: %v", err)
+	}
+	if err := j.dev.WriteBlocks(j.nextBlock, buf.Bytes()); err != nil {
+		return fmt.Errorf("failed to write blocks: %v", err)
+	}
+	// Finally, we'll rescan the journal and make sure that all is well.
+	if err := j.init(); err != nil {
+		return fmt.Errorf("failed to re-scan journal: %v", err)
+	}
+	if got, want := j.current.Revision, e.Revision; got != want {
+		return fmt.Errorf("journal error, latest entry revision %d != written revision %d", got, want)
+	}
+	return nil
+}
+
+// Init scans the journal to figure out the latest valid record, if any.
+func (j *Journal) init() error {
+	// Start where all good stories do: at the beginning!
+	lba := j.start
+	var lastEntry entry
+	nextWriteLBA := j.start
+	for lba < j.start+j.length {
+		br := newBlockReader(j.dev, lba)
+		e, err := unmarshalEntry(br)
+		if err != nil {
+			if lastEntry.Revision > 0 {
+				glog.V(2).Infof("Scanned to invalid entry, using last good entry seen@rev %d (%v)", lastEntry.Revision, err)
+				// We already found the lastet record in the journal, so we're done.
+				break
+			}
+			// Since we haven't already found a good entry, and we were unable
+			// to unmarshal this one,
+			// this means that either:
+			//  a) the journal is completely empty, or
+			//  b) the previously good entry/ies at the start of the journal
+			//     have been completely or partially overwritten during a
+			//     failed write attempt.
+			// Either way, we don't have a valid entry wth a length field we can
+			// rely on, so we we'll have to fall back to scanning all blocks to
+			// look for one.
+			lba++
+			continue
+		}
+		if e.Revision > lastEntry.Revision {
+			glog.V(3).Infof("Scan found rev %d(@ block %d), continuing", e.Revision, lba)
+			// We've found a(nother) good entry, so update our state
+			lastEntry = *e
+			// Skip past the blocks we've just read
+			lba = (br.pos-1)/br.dev.BlockSize() + 1
+			// If this turns out to be the last good entry, then we'll write
+			// at the next block.
+			nextWriteLBA = lba
+			// But loop around again, just in case there are yet more good
+			// entries following on...
+			continue
+		} else if e.Revision < lastEntry.Revision {
+			glog.V(2).Infof("Found newest entry @ rev %d", lastEntry.Revision)
+			// We've found an older revision following a newer one, so we're done.
+			nextWriteLBA = lba
+			break
+		} else {
+			return fmt.Errorf("journal is corrupt - found two entries with the same revision (%d)", e.Revision)
```

Comment:
Or found an entry with revision=0

Reply:
Ah - zero isn't a valid revision (see `Update` - it's always written as `<something>+1`)
I've updated the comment on `Revision` in the `entry` struct to make that clearer, and enforced it in `unmarshalEntry`